java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll

```diff
@@ -30,13 +31,132 @@ private class FollowsSanitizingPrefix extends UnsafeUrlForwardSanitizer {
   FollowsSanitizingPrefix() { this.asExpr() = any(SanitizingPrefix fp).getAnAppendedExpression() }
 }
 
+/** A barrier guard that protects against URL forward vulnerabilities. */
+abstract class UnsafeUrlForwardBarrierGuard extends DataFlow::BarrierGuard { }
+
+/**
+ * Holds if `ma` is a call to a method that checks exact match of string.
+ */
+private predicate isExactStringPathMatch(MethodAccess ma) {
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod().getName() = ["equals", "equalsIgnoreCase"]
+}
+
+/**
+ * Holds if `ma` is a call to a method that checks a path string.
+ */
+private predicate isStringPathMatch(MethodAccess ma) {
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod().getName() =
+    ["contains", "startsWith", "matches", "regionMatches", "indexOf", "lastIndexOf"]
+}
+
+/**
+ * Holds if `ma` is a call to a method of `java.nio.file.Path` that checks a path.
+ */
+private predicate isFilePathMatch(MethodAccess ma) {
+  ma.getMethod().getDeclaringType() instanceof TypePath and
+  ma.getMethod().getName() = "startsWith"
+}
+
+/**
+ * Holds if `ma` protects against path traversal, by either:
+ * * looking for the literal `..`
+ * * performing path normalization
+ */
+private predicate isPathTraversalCheck(MethodAccess ma, Expr checked) {
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod().hasName(["contains", "indexOf"]) and
+  ma.getAnArgument().(CompileTimeConstantExpr).getStringValue() = ".." and
+  ma.(Guard).controls(checked.getBasicBlock(), false)
+  or
+  ma.getMethod() instanceof PathNormalizeMethod and
+  checked = ma
+}
+
+/**
+ * Holds if `ma` protects against double URL encoding, by either:
+ * * looking for the literal `%`
+ * * performing URL decoding
+ */
+private predicate isURLEncodingCheck(MethodAccess ma, Expr checked) {
+  // Search the special character `%` used in url encoding
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod().hasName(["contains", "indexOf"]) and
+  ma.getAnArgument().(CompileTimeConstantExpr).getStringValue() = "%" and
+  ma.(Guard).controls(checked.getBasicBlock(), false)
+  or
+  // Call to `URLDecoder` assuming the implementation handles double encoding correctly
```

Comment:
It does not.

```
jshell> URLDecoder.decode("%25C3%2596", "UTF-8");
$1 ==> "%C3%96"
```

(from https://stackoverflow.com/questions/20679604/does-class-urldecoder-decode-handle-double-encoding)

You don't necessarily have to handle this in this query -- if it's difficult to determine whether they're accounting for double encoding then it might be an acceptable false negative -- but you shouldn't claim you're doing so in the QL comments.

Reply:
Thanks @smowton for reviewing this PR. Shall I just remove the comment since we cannot assume double encoding is handled correctly in the QL comments? Or shall we remove this check completely? I can also add a check to find out whether URLDecoder is used in a loop just like the example https://github.com/github/codeql/blob/877c52981ff3b333d13cdb802b7f3592e5d753bc/java/ql/test/experimental/query-tests/security/CWE-552/UnsafeServletRequestDispatch.java#L119-L122. Will this suffice? Please advise. Thanks.

Reply:
Let's keep it simple for now; just note that double-encoding is a risk this query doesn't handle and maybe we'll look into that as future work.

Reply:
Thanks @smowton for the advice. We'll keep it simple for now.